### PR TITLE
add validate for advanced audit policy

### PR DESCRIPTION
**Pull request by CaoShuFeng**
This change checks group name and non-resrouce URLs format for audit
policy.

**Release note**:

```
add validate for advanced audit policy, kube-apiserver will do a stricter validation and will break existing users with invalid configs.
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go [hunk 1]
```diff
@@ -17,6 +17,9 @@ limitations under the License.
 package validation
 
 import (
+	"strings"
+
+	genericvalidation "k8s.io/apimachinery/pkg/api/validation"
```
Comment (ericchiang) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go]: Why the import rename?

File: staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go [hunk 2]
```diff
@@ -60,3 +65,38 @@ func validateLevel(level audit.Level, fldPath *field.Path) field.ErrorList {
...
+func validateResources(groupResources []audit.GroupResources, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	for _, groupResource := range groupResources {
+		if len(groupResource.Group) == 0 {
+			continue
+		}
+
+		// an error is returned for group name like rbac.authorization.k8s.io/v1beta1
+		// rbac.authorization.k8s.io is the valid one
+		if msgs := genericvalidation.NameIsDNSSubdomain(groupResource.Group, false); len(msgs) != 0 {
```
Comment (ericchiang) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go]: I don't know if this is the correct way to validate a group name. There should be something to parse a group version such as:

https://godoc.org/k8s.io/apimachinery/pkg/runtime/schema#ParseGroupResource
https://godoc.org/k8s.io/apimachinery/pkg/runtime/schema#ParseGroupVersion

cc @timstclair @sttts any comments here?

File: staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go [hunk 3]
```diff
@@ -60,3 +65,38 @@ func validateLevel(level audit.Level, fldPath *field.Path) field.ErrorList {
...
+			allErrs = append(allErrs, field.Invalid(fldPath, url, "invalid non-resource URL format, only support one tailing * at the end"))
+		}
+	}
+	return allErrs
+}
+
+func validateResources(groupResources []audit.GroupResources, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	for _, groupResource := range groupResources {
+		if len(groupResource.Group) == 0 {
```
Comment (ericchiang) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go]: Why is this being skipped? Can you add a comment?

File: staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go [hunk 4]
```diff
@@ -60,3 +65,38 @@ func validateLevel(level audit.Level, fldPath *field.Path) field.ErrorList {
 		return field.ErrorList{field.NotSupported(fldPath, level, validLevels)}
 	}
 }
+
+func validateNonResourceURLs(urls []string, fldPath *field.Path) field.ErrorList {
```
Comment (ericchiang) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go]: I'm worried that this method is brittle. Might be worth looking for an agreed upon validator somewhere in the rest of the codebase. e.g. What does RBAC do to validate those fields?

**Review by ericchiang — COMMENTED** [review]
Thanks for this!

Would be worth a release note that this is getting stricter since this will break existing users with invalid configs.

File: staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go [hunk 5]
```diff
@@ -60,3 +65,38 @@ func validateLevel(level audit.Level, fldPath *field.Path) field.ErrorList {
...
+	var allErrs field.ErrorList
+	for _, url := range urls {
+		if url == "*" {
+			continue
+		}
+
+		if !strings.HasPrefix(url, "/") {
+			allErrs = append(allErrs, field.Invalid(fldPath, url, "invalid non-resource URL format"))
```
Comment (timstclair) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go]: nit: should be `fldPath.Index(i)`, same below (need `i` from the `for`)

File: staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go [hunk 6]
```diff
@@ -60,3 +65,38 @@ func validateLevel(level audit.Level, fldPath *field.Path) field.ErrorList {
...
+		}
+
+		count := strings.Count(url, "*")
+		if count > 1 || (count == 1 && !strings.HasSuffix(url, "*")) {
```
Comment (timstclair) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go]: nit: a better way of expressing this would be
```go
if !strings.ContainsRune(url[:len(url) - 1]) {
```

**Comment by ericchiang** [comment]
@k8s-bot ok to test

/unassign @lavalamp @timothysc

/assign @ericchiang @timstclair

File: staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go (hunk 1 above)
Comment (CaoShuFeng) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go]: This package is also named validation, rename the imported one to distinguish them.
Do we have some convention for this?

File: staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go (hunk 4 above)
Comment (CaoShuFeng) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go]: liggitt saies that RBAC doesn't do validation in kube-apiserver.
But I fount this:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/audit/policy/checker.go#L112

File: staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go (hunk 2 above)
Comment (CaoShuFeng) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go]: https://github.com/kubernetes/community/blame/master/contributors/devel/api-conventions.md#L92
This document defines a group name format.

File: staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go (hunk 5 above)
Comment (CaoShuFeng) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go]: Done.

File: staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go (hunk 6 above)
Comment (CaoShuFeng) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go]: Done.

File: staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go (hunk 3 above)
Comment (CaoShuFeng) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go]: Done.

**Comment by CaoShuFeng** [comment]
These 2 tests fail because some CI use invalid rule configs.
I will work on it.
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/47784/pull-kubernetes-kubemark-e2e-gce/36469/artifacts/e2e-36469-master/kube-apiserver.log

File: staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go (hunk 1 above)
Comment (ericchiang) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go]: You can't refer to the package within itself, so `validation.foo` already unambiguously refers to the imported package

File: staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go (hunk 2 above)
Comment (ericchiang) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go]: Oh great. Can you link to that in the comment? And note that "Group names must be lower case and be valid DNS subdomains."

File: cluster/gce/gci/configure-helper.sh [hunk 7]
```diff
@@ -493,18 +493,18 @@ function create-master-audit-policy {
   # Known api groups
   local -r known_apis='
       - group: "" # core
-      - group: "admissionregistration.k8s.io/v1alpha1"
-      - group: "apps/v1beta1"
+      - group: "admissionregistration.k8s.io"
```
Comment (timstclair) [inline cluster/gce/gci/configure-helper.sh]: Oops! I'm going to send this as a separate PR, because we might try and cherrypick it to 1.7

Comment (CaoShuFeng) [inline cluster/gce/gci/configure-helper.sh]: @timstclair 
Got it, thanks.

File: staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go (hunk 1 above)
Comment (CaoShuFeng) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go]: Done.

File: staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go (hunk 2 above)
Comment (CaoShuFeng) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go]: Done.

**Comment by CaoShuFeng** [comment]
@ericchiang Thanks for your review.
Please take a look again.

File: staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go [hunk 8]
```diff
@@ -60,3 +65,40 @@ func validateLevel(level audit.Level, fldPath *field.Path) field.ErrorList {
...
+		if url == "*" {
+			continue
+		}
+
+		if !strings.HasPrefix(url, "/") {
+			allErrs = append(allErrs, field.Invalid(fldPath.Index(i), url, "invalid non-resource URL format"))
+		}
+
+		if strings.ContainsRune(url[:len(url)-1], '*') {
+			allErrs = append(allErrs, field.Invalid(fldPath.Index(i), url, "invalid non-resource URL format, only support one trailing * at the end"))
```
Comment (ericchiang) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go]: nit: to keep this error consistent with the wording of other validation rules maybe "non-resource URL wildcards '*' must be the final character of the rule"

File: staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go [hunk 9]
```diff
@@ -60,3 +65,40 @@ func validateLevel(level audit.Level, fldPath *field.Path) field.ErrorList {
...
+
+func validateNonResourceURLs(urls []string, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	for i, url := range urls {
+		if url == "*" {
+			continue
+		}
+
+		if !strings.HasPrefix(url, "/") {
+			allErrs = append(allErrs, field.Invalid(fldPath.Index(i), url, "invalid non-resource URL format"))
```
Comment (ericchiang) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go]: nit: need to have the error explain the validation. e.g. "non-resource URL rules must begin with a '/' character"

**Review by ericchiang — COMMENTED** [review]
couple nits on the error messages. other then that lgtm. thanks!

File: staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go (hunk 9 above)
Comment (CaoShuFeng) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go]: Done.

File: staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go (hunk 8 above)
Comment (CaoShuFeng) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go]: Done.

File: staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go [hunk 10]
```diff
@@ -60,3 +65,40 @@ func validateLevel(level audit.Level, fldPath *field.Path) field.ErrorList {
...
+	for i, url := range urls {
+		if url == "*" {
+			continue
+		}
+
+		if !strings.HasPrefix(url, "/") {
+			allErrs = append(allErrs, field.Invalid(fldPath.Index(i), url, "non-resource URL rules must begin with a '/' character"))
+		}
+
+		if strings.ContainsRune(url[:len(url)-1], '*') {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go]: can url be empty?

Comment (CaoShuFeng) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go]: An empty url matches nothing:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/audit/policy/checker.go#L112
I think an empty url doesn't make any sense, so empty url is invalid for me.

Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go]: Right, I meant that `url[:len(url)-1]` will panic then, if I am not wrong.

Comment (CaoShuFeng) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/validation/validation.go]: Done.
I also add an empty string in unit test. An empty string will cause an error in line 76.
